pkg/secretless/entrypoint/start_secretless.go

```diff
@@ -0,0 +1,125 @@
+package entrypoint
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+
+	"github.com/cyberark/secretless-broker/internal"
+	secretlessLog "github.com/cyberark/secretless-broker/internal/log"
+	"github.com/cyberark/secretless-broker/internal/plugin/v1/event_notifier"
+	"github.com/cyberark/secretless-broker/pkg/secretless"
+	"github.com/cyberark/secretless-broker/pkg/secretless/config"
+	v2 "github.com/cyberark/secretless-broker/pkg/secretless/config/v2"
+
+	"github.com/cyberark/secretless-broker/internal/plugin"
+	"github.com/cyberark/secretless-broker/internal/profile"
+	"github.com/cyberark/secretless-broker/internal/signal"
+)
+
+// CLIParams holds the command line flag information that Secretless was started
+// with.
+type CLIParams struct {
+	ConfigFile          string
+	ConfigManagerSpec   string
+	DebugEnabled        bool
+	FsWatchEnabled      bool
+	PluginChecksumsFile string
+	PluginDir           string
+	ProfilingMode       string
+	ShowVersion         bool
+}
+
+// StartSecretless method is the main entry point into the broker after the CLI
+// flags have been parsed
+func StartSecretless(params *CLIParams) {
+
+	ShowVersion(params.ShowVersion)
+
+	VerifyPlugins(params.PluginDir, params.PluginChecksumsFile)
+
+	// Construct the deps of Secretless
+	cfg := ReadConfig(params.ConfigFile)
+	logger := secretlessLog.New(params.DebugEnabled)
+	evtNotifier := event_notifier.New(nil)
+	availPlugins := &internal.AvailPluginStub{}
+
+	// Prepare Secretless
+	secretless := internal.NewSecretless(cfg, availPlugins, logger, evtNotifier)
+	signal.StopOnExitSignal(secretless)
+
+	StartPerformanceProfiling(params.ProfilingMode)
+
+	secretless.Start()
+}
+
+func ReadConfig(cfgFile string) v2.Config {
+	// TODO: Add back in CRD / generalized config option
+	cfg, err := config.LoadFromFile(cfgFile)
+	if err != nil {
+		log.Fatalln(err)
+	}
+	return cfg
+}
+
+func ShowVersion(showVersionAndExit bool) {
+	if showVersionAndExit {
+		fmt.Printf("secretless-broker v%s\n", secretless.FullVersionName)
+		os.Exit(0)
+	}
+	log.Printf("Secretless v%s starting up...", secretless.FullVersionName)
+}
+
+// StartPerformanceProfiling starts a perfomance profiling, and sets up an
+// os.Signal listener that will automatically call Stop() on the profile
+// when an system halt is raised.
+func StartPerformanceProfiling(profileType string) {
+	// No profiling was requested
+	if profileType == "" {
+		return
+	}
+
+	// Validate requested type
+	if err := profile.ValidateType(profileType); err != nil {
+		log.Fatalln(err)
```

Comment:
There's a few places where `log.Fatalln` is used to terminate the program "relatively deep in the stack". Any reason to not propagate errors upwards so that these methods don't have this catastrophic side-effect ?

Reply:
Well, the idea is that `StartSecretless` is, in fact, the highest level, logically speaking.  There no one to handle errors up here.  `main` just parses cmd line flags and delegates to `StartSecretless`.

That said, we _probably_ want to make `StartSecretless` a proper function object that we can unit test at some point, and if we did that I would definitely do what you're suggesting.  I sort of feel like that should be a separate story though.

Reply:
Yup, definitely coming from the unit testable angle. Happy for it to be a separate story.